File: C++/String/6_count_and_say.cpp
// Online C++ compiler to run C++ program online
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    string countAndSay(int n)
    {
        if (n == 1)
            return "1";

        string s = "1";
        for (int i = 2; i <= n; ++i)
        {
            string temp = "";
            int count = 1, prev = s[0] - '0'; // Convert char to int

            for (int j = 1; j < s.size(); ++j)
            {
                if (prev == s[j] - '0') // Convert char to int
                    count++;
                else
                {
                    temp += to_string(count);
                    temp += to_string(prev);
                    count = 1; // Reset count
                }

                prev = s[j] - '0';
            }
            temp += to_string(count); // Append last count
            temp += to_string(prev);  // Append last prev
            s = temp;
        }
        return s;
    }
};



int main() {
    // Write C++ code here
    Solution ob;
    cout << ob.countAndSay(2);
}

File: Python/String/6_count_and_say.py
"""
Idea:
Iterating from 1 to 'n' , with each iteration making a new string by using current string elements.

"""


class Solution:
    def countAndSay(self, n: int) -> str:
        if n == 1:
            return "1"

        s = "1"
        for i in range(2, n + 1):
            # with each iteration, counting & saying the elements
            temp = ""
            count = 1
            prev = int(s[0])  # Convert char to int

            for j in range(1, len(s)):
                if prev == int(s[j]):
                    count += 1
                else:
                    temp += str(count)
                    temp += str(prev)
                    count = 1  # Reset count

                prev = int(s[j])

            temp += str(count)  # Append last count
            temp += str(prev)  # Append last prev
            s = temp

        return s


if __name__ == "__main__":
    ob = Solution()
    print(ob.countAndSay(4))
